core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -378,6 +421,17 @@ object Cause extends Serializable {
     }
   }
 
+  final case class Data private[Cause] (stackless: Boolean)
```

Comment:
I wonder if this is enough for binary backward compatibility (I don't think so, but I am not sure). You might want to make the `Data` class private too

Reply:
I tried to do that initially but ran into issues because `Data` is exposed when users pattern match against `Meta`. But maybe to your point above if we disallow that then we can make this completely private. We just need to make sure we give users enough functionality to work with `Cause` like in the `ZStream` snippet I updated..

Reply:
Did anyone say streams? 🤪

I'd be happy to move that snippet into `Cause`'s companion object.

Reply:
@iravid Awesome! 😃

Reply:
@adamgfraser Good point. I'm not exactly sure what to do here, but I do know if we don't hide enough, we're gonna have real trouble changing functionality in 1.x. We'd need MiMa to test to for sure either way...

Reply:
Why don't I try make all the constructors private and implementing a new version of `fold`. Then we should be able to make `Data` private. I can also try to set up MiMa so we can test binary compatibility and then we can just turn it on when we go live with 1.0.

Reply:
That would be fantastic! Thank you! 🙏

Reply:
I tried to add MiMa but it looks like it needs to compare a current version against a released version. However, I was able to make the constructors and `Data` private so I think we should be in good shape from a binary compatibility perspective.